parser/src/error.rs

```diff
@@ -0,0 +1,140 @@
+use std::{fmt::Display, iter};
+
+use lalrpop_util::lexer::Token;
+use thiserror::Error;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Location {
+    pub line: usize,
+    pub column: usize,
+    pub absolute: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct Span {
+    pub start: Option<Location>,
+    pub end: Option<Location>,
+}
+
+impl Display for Span {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match (&self.start, &self.end) {
+            (Some(start), Some(end)) if start == end => {
+                write!(f, "[{}:{}]", start.line, start.column)
+            }
+            (Some(location), None) | (None, Some(location)) => {
+                write!(f, "[{}:{}]", location.line, location.column)
+            }
+            (Some(start), Some(end)) => write!(
+                f,
+                "[{}:{}]->[{}:{}]",
+                start.line, start.column, end.line, end.column
+            ),
+            (None, None) => write!(f, "?",),
```

Comment:
I think this
```
Error parsing: <message> at ?
```
would be better than
```
Error parsing: <message> at
```

What do you think?

Reply:
Agreed